// @libp2p/webrtc/dist/src/webrtc/rtc-ice-candidate.js
/**
 * @see https://developer.mozilla.org/docs/Web/API/RTCIceCandidate
 */
export class IceCandidate {
    address;
    candidate;
    component;
    foundation;
    port;
    priority;
    protocol;
    relatedAddress;
    relatedPort;
    sdpMLineIndex;
    sdpMid;
    tcpType;
    type;
    usernameFragment;
    constructor(init) {
        if (init.candidate == null) {
            throw new DOMException('candidate must be specified');
        }
        this.candidate = init.candidate;
        this.sdpMLineIndex = init.sdpMLineIndex ?? null;
        this.sdpMid = init.sdpMid ?? null;
        this.usernameFragment = init.usernameFragment ?? null;
        this.address = null;
        this.component = null;
        this.foundation = null;
        this.port = null;
        this.priority = null;
        this.protocol = null;
        this.relatedAddress = null;
        this.relatedPort = null;
        this.tcpType = null;
        this.type = null;
    }
    toJSON() {
        return {
            candidate: this.candidate,
            sdpMLineIndex: this.sdpMLineIndex,
            sdpMid: this.sdpMid,
            usernameFragment: this.usernameFragment
        };
    }
}
